puma/metrics.py

```diff
@@ -5,11 +5,45 @@
 from puma.utils.histogram import hist_w_unc, save_divide
 
 
+def weighted_percentile(
+    data: np.ndarray,
+    perc: np.ndarray,
+    weights: np.ndarray = None,
+):
+    """Calculate weighted percentile.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        Data array
+    perc : np.ndarray
+        Percentile array
+    weights : np.ndarray
+        Weights array, by default None
+
+    Returns
+    -------
+    np.ndarray
+        Weighted percentile array
+    """
+    if weights is None:
+        weights = np.ones_like(data)
+    ix = np.argsort(data)
+    data = data[ix]  # sort data
+    weights = weights[ix]  # sort weights
+    cdf = np.cumsum(weights) - 0.5 * weights
```

Comment:
why do you halve the weights here?

Reply:
For the possibility to weight also signal events. If they have weights, we treat each of them not as `1` but as its corresponding weight and then calculate the percentile. I added links to Wiki and Stackoverflow.

Reply:
Thanks for the links!